File: cpp/NoiseBuffer2D.cpp
#include "NoiseBuffer2D.h"

NoiseBuffer2D::NoiseBuffer2D(uint32_t width, uint32_t height)
{
    m_width = width;
    m_height = height;
    m_capacity = width * height;
    m_buffer = new double[m_capacity];
}

NoiseBuffer2D::~NoiseBuffer2D()
{
    delete m_buffer;
}

uint32_t NoiseBuffer2D::width() const
{
    return m_width;
}

uint32_t NoiseBuffer2D::height() const
{
    return m_height;
}

double NoiseBuffer2D::at(int row, int col)
{
    assert(0 <= row && row < (int)m_height);
    assert(0 <= col && col < (int)m_width);
    return m_buffer[(((m_height - 1) - row) * m_width) + col];
}

void NoiseBuffer2D::set(int row, int col, double value)
{
    assert(0 <= row && row < (int)m_height);
    assert(0 <= col && col < (int)m_width);
    m_buffer[(((m_height-1) - row) * m_width) + col] = value;
}

void NoiseBuffer2D::fill(double value)
{
    for(uint32_t index = 0; index < m_capacity; ++index)
        m_buffer[index] = value;
}

void NoiseBuffer2D::normalize()
{
    double min = m_buffer[0];
    double max = min;
    for(int index = 1; index < (int)m_capacity; ++index) {
        double value = m_buffer[index];
        if(value < min) min = value;
        if(value > max) max = value;
    }

    (*this) -= min;
    (*this) /= (max - min);
}

void NoiseBuffer2D::operator+=(const NoiseBuffer2D &other)
{
    assert(m_width == other.m_width);
    assert(m_height == other.m_height);
    for(uint32_t index = 0; index < m_capacity; ++index)
        m_buffer[index] += other.m_buffer[index];
}

void NoiseBuffer2D::operator+=(double value)
{
    for(uint32_t index = 0; index < m_capacity; ++index)
        m_buffer[index] += value;
}

void NoiseBuffer2D::operator-=(double value)
{
    for(uint32_t index = 0; index < m_capacity; ++index)
        m_buffer[index] -= value;
}

void NoiseBuffer2D::operator*=(double factor)
{
    for(uint32_t index = 0; index < m_capacity; ++index)
        m_buffer[index] *= factor;
}

void NoiseBuffer2D::operator/=(double factor)
{
    for(uint32_t index = 0; index < m_capacity; ++index)
        m_buffer[index] /= factor;
}

File: py/NoiseBuffer2D.py
class NoiseBuffer2D:

    def __init__(self, width, height):
        self._width = width
        self._height = height
        self._capacity = width * height
        self._buffer = [0.0] * self._capacity

    def width(self):
        return self._width

    def height(self):
        return self._height

    def at(self, row, col):
        assert(0 <= row and row < int(self._height))
        assert(0 <= col and col < int(self._width))
        return self._buffer[(((self._height - 1) - int(row)) * self._width) + int(col)]

    def set(self, row, col, value):
        assert(0 <= row and row < int(self._height))
        assert(0 <= col and col < int(self._width))
        self._buffer[(((self._height-1) - int(row)) * self._width) + int(col)] = value

    def fill(self, value):
        for index in range(self._capacity):
            self._buffer[index] = value

    def normalize(self):
        low = min(self._buffer)
        high = max(self._buffer)

        self -= low
        self /= (high - low)

    def __iadd__(self, other):
        # += Operator
        if isinstance(other, NoiseBuffer2D):
            assert(self._width == other.self._width)
            assert(self._height == other.self._height)
            for index in range(self._capacity):
                self._buffer[index] += other.self._buffer[index]
        else:
            assert(isinstance(other, (int, float)))
            for index in range(self._capacity):
                self._buffer[index] += other
        return self

    def __isub__(self, other):
        # -= Operator
        if isinstance(other, NoiseBuffer2D):
            assert(self._width == other.self._width)
            assert(self._height == other.self._height)
            for index in range(self._capacity):
                self._buffer[index] -= other.self._buffer[index]
        else:
            assert(isinstance(other, (int, float)))
            for index in range(self._capacity):
                self._buffer[index] -= other
        return self

    def __imul__(self, factor):
        # *= Operator
        for index in range(self._capacity):
            self._buffer[index] *= factor
        return self

    def __itruediv__(self, factor):
        # /= Operator
        for index in range(self._capacity):
            self._buffer[index] /= factor
        return self
